en/developer/customization/articles/50-liferay-forms/02-form-storage-adapters.markdown

```diff
@@ -39,61 +36,166 @@ their creation, and a different storage adapter cannot be selected.
+    request.
+
+- Method: `public DDMForm getDDMForm()`
+    Returns the `DDMForm` object of the form the record is associated with, for
+    deletion.
+
+`DDMStorageAdapterGetRequest`:
+
+- Method: `public long getPrimaryKey()`
+    Returns the `storageId` <!-- ???--> of the form record, which is useful for
+    any work that requires identifying a record.
+
+## Implementing a `DDMStorageAdapter`
+
+Here's an overview of what you must accomplish in implementing each storage
+adapter method:
+
+`save`
+: Detect whether the form record is new or if it already exists (due to the
+auto-save feature). If new, call the DDM service for adding `DDMContent` to the
+database, passing to it, among other things, the serialized version (transformed
+into your storage format) of the `DDMFormValues`.
+
+`delete`
+: There's no serialization or deserialization involved with deletion. If you
+don't need additional custom logic for deletion, you can use the default storage
+adapter's basic deletion code, which calls the DDM content service's `delete`
+method, passing the primary key of the form record, as retrieved from the delete
+request. This will be demonstrated in the next tutorial. 
+
+`get`
+: Get the DDM content using the storage ID of the form record, retrieved from
+the request object, then construct the `DDMFormValues` object from the
+serialized data you've already stored, by calling the deserialization code. Set
+the now-deserialized `DDMFormValues` into the response object. Of course, if you
+need to add custom logic in the meantime, do that too.
+
+Serializing and deserializing the form record's content requires some additional
+work and documentation for the interfaces Liferay provides will be written. 
+
+The basic idea is that when you're reading `DDMContent`, it must be deserialized
+from the storage format back into a `DDMFormValues` object that the
+`DDMStorageAdapterGetResponse`'s `Builder` constructor requires. Conversely,
+when you're saving `DDMcontent`, you need to get the `DDMFormValues` passed to
+the get request, and serialize it into the storage format. There's a
+`DDMFormValuesSerializer` to implement for the serialization logic, and a
+`DDMFormValuesDeserializer` to implement for the deserialization logic.
+Liferay's default implementations, which support transforming the
+`DDMFormValues` object to JSON and vice versa, can be found
+[here](https://github.com/liferay/liferay-portal/blob/7.2.x/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/io/DDMFormValuesJSONSerializer.java)
+and
+[here](https://github.com/liferay/liferay-portal/blob/7.2.x/modules/apps/dynamic-data-mapping/dynamic-data-mapping-service/src/main/java/com/liferay/dynamic/data/mapping/internal/io/DDMFormValuesJSONDeserializer.java),
+respectively.
+
+<!-- unnecessary now? Looks like we have a storage manager impl that calls
```

Comment:
@natocesarrego when looking at the code I determined we don't need to do validation in the storage adapter anymore. Is that correct? Just want to confirm.

Reply:
Yes! Please, go ahead and delete this validation section. 😉